beacon-chain/casper/validator.go

```diff
@@ -184,3 +184,95 @@ func TotalActiveValidatorDepositInEth(validators []*pb.ValidatorRecord) uint64 {
 
 	return depositInEth
 }
+
+// AddValidator runs for every validator that is inducted as part of a log created on the PoW chain.
+func AddValidator(
+	validators []*pb.ValidatorRecord,
+	pubKey []byte,
+	withdrawalShard uint64,
+	withdrawalAddr []byte,
+	randaoCommitment []byte) []*pb.ValidatorRecord {
+
+	// TODO(#633): Use BLS to verify signature proof of possession and pubkey and hash of pubkey.
+
+	newValidatorRecord := &pb.ValidatorRecord{
+		Pubkey:            pubKey,
+		WithdrawalShard:   withdrawalShard,
+		WithdrawalAddress: withdrawalAddr,
+		RandaoCommitment:  randaoCommitment,
+		Balance:           uint64(params.GetConfig().DepositSize * params.GetConfig().Gwei),
+		Status:            uint64(params.PendingActivation),
+		ExitSlot:          0,
+	}
+
+	index := minEmptyValidator(validators)
+	if index > 0 {
+		validators[index] = newValidatorRecord
+		return validators
+	}
+
+	validators = append(validators, newValidatorRecord)
+	return validators
+}
+
+// ChangeValidators updates the validator set during state transition.
+func ChangeValidators(currentSlot uint64, totalPenalties uint64, validators []*pb.ValidatorRecord) []*pb.ValidatorRecord {
+	maxAllowableChange := uint64(2 * params.GetConfig().DepositSize * params.GetConfig().Gwei)
+
+	totalBalance := TotalActiveValidatorDeposit(validators)
+
+	// Determine the max total wei that can deposit and withdraw.
+	if totalBalance > maxAllowableChange {
+		maxAllowableChange = totalBalance
+	}
+
+	var totalChanged uint64
+	for i := 0; i < len(validators); i++ {
+		if validators[i].Status == uint64(params.PendingActivation) {
+			validators[i].Status = uint64(params.Active)
+			totalChanged += uint64(params.GetConfig().DepositSize * params.GetConfig().Gwei)
+			// TODO(#614): Add validator set change.
+		}
+		if validators[i].Status == uint64(params.PendingExit) {
+			validators[i].Status = uint64(params.PendingWithdraw)
+			validators[i].ExitSlot = currentSlot
+			totalChanged += validators[i].Balance
+			// TODO(#614): Add validator set change.
+		}
+		if totalChanged > maxAllowableChange {
+			break
```

Comment:
Shouldnt this break before the validator balance is updated ?

Reply:
I don't think it matters, it's either going to break in the current loop or before the next loop starts 😅

Reply:
I think it matters since we are setting the 
`validators[i].Status = uint64(params.Active)`
and 
`validators[i].ExitSlot = currentSlot` 
if the `totalChanged > maxAllowableChange` then the validator's status and exit slot should not be updated.
imo the check should come in before we change any part of the validator record

Reply:
I mean it's going to break in the current loop regardless. I just confirmed with Danny on gitter as well:

>We stop if and after we process a validator that sends us over the max allowable change (or equals it).
thus it is after the two if statements
when we bring a validator in or out, we make note of how much we have changed the total active validator set balance by. If we process a validator and then are over the limit, we don’t do any more for this state update

Reply:
Alright in that case it sounds good to me then